add support for shell completion

WillAbides opened the pull request:
This adds hooks for supporting shell completion but doesn't contain any actual shell completion code.  I don't believe there are any breaking changes in this PR and any changes to existing behavior will be a result of modifying `kong.CompletionOptions`

I haven't made any documentation changes yet.  I want to get some feedback on this before putting in that work.

`kong.CompletionOptions` / `kong.ConfigureCompletion`<sup>[1](#foot1)</sup> are the only API additions.  

### The options

#### Completer

Completer is where most of the action is.  This is the function that actually provides the completions.  If you want to implement it with for a basic case, you could set it to `kongplete.Complete` <sup>[2](#foot2)</sup>. For custom options it would be something like :

```golang
func(ctx *kong.Context) (bool, error) {
	return kongplete.Complete(ctx,
		kongplete.WithPredictor("mypredictor", complete.PredictSet("foo", "bar"), 
	)
}
```

The completer is responsible for determining whether this is a completion run or not.  Normally that will be by checking for the presence of the environment variable `COMP_LINE`.  I didn't want to move that into kong because that would make kong responsible for any differences between shells.

The completer is also responsible for writing the completion options to stdout.  This should be done by writing to `ctx.Stdout`. Another option would have been to make the Complete interface return a list of options for kong to write. I don't have a stong opinion on this one. I slightly prefer leaving it up to the completer because that allows more flexibility.

#### CompletionInstaller

CompletionInstaller provides the functions to install completion to the shell (edit .bash_profile, zsh or whatnot).  The functions are pretty straight forward.  They should do whatever configuration is necessary and return nil, or an error depending on success.  They are implemented similar to help, so the program is terminated immediately after they run.  By default their flags are `--install-shell-completion` and `--uninstall-shell-completion`

#### All the rest

The rest of the options control the appearance of the install and uninstall flags in help.  They are self-explanatory (I hope)

### Finally

As mentioned above, this is a bare-bones PR.  It could be merged as-is, with documentation referring to kongplete for implementation.  Another option would be to move kongplete into the kong repo.  A third option would be providing a Completer in the kong repo that pulls in kongplete.  I think the best usability is the middle option (moving kongplete to the kong repo).  However that would require importing `github.com/posener/complete`.  I see that kong currently has only one external import, and I suspect that is an intentional decision.

I'm looking forward to feedback.

---

<a name="foot1">1</a>: I'm not sure how `kong.ConfigureCompletion` adds value.  I was copying from `kong.HelpOptions` and added the equivalent to `kong.ConfigureHelp` for completeness. 

<a name="foot2">2</a>: I haven't actually updated kongplete to match this interface yet, so these examples are aspirational.

alecthomas commented:
> I think the best usability is the middle option (moving kongplete to the kong repo).

Yes, agreed, I think this is the best option. I'd prefer to have comprehensive completion included in Kong itself without users having to think about it too much.

>  However that would require importing github.com/posener/complete. I see that kong currently has only one external import, and I suspect that is an intentional decision.

This is not specifically intentional, it just happened to be the case :)

alecthomas commented on completion.go:
```diff
@@ -0,0 +1,145 @@
+	UninstallFlagHelp string
+
+	// whether to hide --install-shell-completion and --uninstall-shell-completion from help.  default: false
+	HideFlags bool
```
I'd prefer the flags to just be separate types (as you have below with `uninstallCompletionsValue`), then the users can decide how they want to use them as they would any other flag. It will also reduce the complexity of this config struct.

WillAbides replied:
I think you are saying that we can get rid of `HideFlags` all the `*Name` and `*Help` fields above and replace them with exported types that people can put in their own cli structs with whatever names, help messages or other attributes they see fit.  If so, that sounds good to me.

alecthomas commented on completion.go:
```diff
@@ -0,0 +1,145 @@
+
+// Completer is a function to run shell completions. Returns true if this was a completion run. Kong will exit 0
+// immediately when it returns true.
+type Completer func(*Context) (bool, error)
```
Shouldn't this accept positional information of some sort? Or I guess that is part of the remaining integration in `kongcomplete`?

WillAbides replied:
That is handled by [PositionalPredictor](https://github.com/WillAbides/kongplete/blob/19e042f32715fd238070f8f8e3c23ad6024ed8ef/internal/positionalpredictor/positional.go) in kongplete.

alecthomas replied:
This type needs documentation on how it conveys completion information. It looks, from the tests, like it does this via stdout, but the comments need to be explicit about how and what is expected.

alecthomas commented on completion.go:
```diff
@@ -0,0 +1,145 @@
+//CompletionInstaller contains functions to install completions from to a shell.
+type CompletionInstaller interface {
+	Install(ctx *Context) error
+	Uninstall(ctx *Context) error
```
What do implementations of this interface actually do, write completion scripts to stdout?

WillAbides replied:
They will detect the available shell and edit the appropriate config file.  For example, in bash, it will add a line like this to `.bash_profile`:

```shell
complete -C /path/to/bin/myprogram myprogram
``` 

The implementation in posener/complete handles bash, fish and zsh.

WillAbides commented:
Thanks for the feedback.  I will continue by making the configuration changes you requested and moving kongplete into this PR.

Once kongplete is part of kong, I think it makes sense to merge kongplete's Options into kong's. We can get rid of all of the kongplete options aside from `WithPredictors` (which should probably be renamed to `Predictors` to fit with kong's options). The rest of the options primarily existed to test kongplete separately from kong.

WillAbides commented:
doh. I pushed 122ad6c9eb0d9a2910c3fec0bac8983a9744d10c too soon.  That won't work.  Correction coming.

WillAbides commented:
As of d14caa7316e161ffeeb254c3b33a25aadda16c51, setting up `install-completion` and `uninstall-completion` will look like this:  

```golang
var myCli struct {
	DoIt   struct{} `kong:"cmd"`
	Config struct {
		InstallCompletion   kong.InstallCompletion   `kong:"cmd,help='Install shell completion.'"`
		UninstallCompletion kong.UninstallCompletion `kong:"cmd,help='Uninstall shell completion.'"`
	}
}
```

I like it better as a command than the flag that I had before because you can control where it shows up (in the case above, under a config sub-command).

WillAbides commented:
The lions' share is done.  Documentation needs to be updates, and there's probably some cleanup to do.

alecthomas commented on predictors.go:
```diff
@@ -0,0 +1,240 @@
+	// If the last character in the command line is space, this would be the
+	// last word, otherwise, it would be the word before that.
+	LastCompleted string
+}
```
Does this capture intra-token completions? eg. `token` where the cursor is at `tok><en`?

WillAbides replied:
Yes.  If you had typed `cmdname this is my tok<tab>`, PredictorArgs would look like this:

```go
PredictorArgs{
	All:           []string{"this", "is", "my", "tok"},
	Completed:     []string{"this", "is", "my"},
	Last:          "tok",
	LastCompleted: "my",
}
```

Likewise, if you had typed `cmdname this is my token t<4 x back arrow><tab>` (putting the cursor after the k in token), PredictorArgs would look the same as above.

WillAbides commented on completion_test.go:
```diff
@@ -0,0 +1,351 @@
+		} `kong:"cmd"`
+	}
+
+	type completeTest struct {
```
I added a few more tests to cover mid-token and backarrowed commands.

WillAbides commented:
@alecthomas I addressed your comments from the most recent review. Next I'm going to work on README.md to make sure this is something that can conveyed simply.

One thing I don't like about the interface is the use of the word "predictor". It works in `posener/complete` because the whole package is dealing with completions and it is obvious what would be predicted. In kong, the meaning is less obvious. "completion predictor" would put it in context, but that's getting a bit long.  Any suggestions are appreciated here.

alecthomas commented on completion.go:
```diff
@@ -0,0 +1,177 @@
+type CompletionOptions struct {
+	// RunCompletion determines whether to respond to shell completion requests. default: false
+	// When RunCompletion is false, all other options in CompletionOptions are ignored.
+	RunCompletion bool
```
Unless there's some downside I'm unaware of, I'd prefer this to either be removed altogether or default to true.

alecthomas replied:
Unless there's some downside I'm unaware of, I'd prefer this to either be removed altogether or default to true.

alecthomas commented on completion.go:
```diff
@@ -0,0 +1,177 @@
+
+	// Completer is the function to run for shell completions. If nil, a default completer based on github.com/posener/complete
+	// will be run.
+	Completer Completer
```
Can you document in the comments what exactly this function should do?

alecthomas commented on completion.go:
```diff
@@ -0,0 +1,177 @@
+	if inst == nil {
+		inst = NewCompletionInstaller(nil, nil)
+	}
+	return inst.Install(ctx)
```
I tested this with `_examples/docker` and there's no indication from posener/complete as to what it is doing. Can that package be configured to provide something useful to the user?

Additionally, the errors from posener/complete are really ugly (I think due to the use of multi-err). eg. when attempting to install completion twice:

```
docker: error: 1 error occurred:
               	* already installed in /Users/aat/.zshrc
 ```

I'm wondering if we can make these errors not so ugly :\

alecthomas commented on options.go:
```diff
@@ -174,6 +174,14 @@ func ConfigureHelp(options HelpOptions) Option {
 	})
 }
 
+// ConfigureCompletion sets options for shell completion.
```
Don't worry about having this function, just rely on the method on CompletionOptions. The only reason it exists for other things is legacy.

alecthomas commented on predictors.go:
```diff
@@ -0,0 +1,240 @@
+	return predictorFunc(func(args PredictorArgs) []string {
+		return options
+	})
+}
```
Please add predictors for the builtin types `file` `existingfile` and `path`, as well as `enum` (if not already).

This would be a) very useful and b) serve as a good example of how to write predictors that do more than return a static set.
